### PR TITLE
Updated the SystemFileStream class to always own the underlying SystemFile

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
The SystemFileStream has also been simplified. The fake length and base offset to seek within the SystemFile has been removed and a constructor which accepts a path and an open mode which mirrors the Open() function.

The OpenMode enum class and utility functions has been moved to a OpenMode.h header.

Signed-off-by: lumberyard-employee-dm <56135373+lumberyard-employee-dm@users.noreply.github.com>

## How was this PR tested?

Ran AzCore.Tests and built all code which created a SystemFilesStream

### CODE REVIEW

**Review by rgba16f — APPROVED** [review]

**Review by gadams3 — APPROVED** [review]

**Comment by gadams3** [comment]
This looks much better, cleaner, easier to use, especially with the multiple constructors, managing everything on construction and destruction.

File: Code/Framework/AzCore/AzCore/IO/GenericStreams.h [hunk 1]
```diff
@@ -61,38 +62,47 @@ namespace AZ::IO
...
-        SizeType    GetCurPos() const override { return m_curPos; }
+        SizeType    GetCurPos() const override;
         SizeType    GetLength() const override;
         const char* GetFilename() const override;
-        OpenMode    GetModeFlags() const override { return m_mode; }
+        OpenMode    GetModeFlags() const override;
         bool        ReOpen() override;
 
+        //! Moves a SystemFile instance from an rvalue or xvalue SystemFileStream
+        SystemFile MoveSystemFile() &&;
```
Comment (amzn-phist) [inline Code/Framework/AzCore/AzCore/IO/GenericStreams.h:101]: Is this function used anywhere?  I didn't see it.

What is this syntax called btw?

**Review by amzn-phist — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/IO/GenericStreams.h (hunk 1 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/IO/GenericStreams.h:101]: This function is not used anywhere yet.

This syntax here is a [reference qualifier ](https://en.cppreference.com/w/cpp/language/member_functions#member_functions_with_ref-qualifier)

The purpose of this function is allow the use to move or "steal" the SystemFile from an expiring `SystemFileStream`
ex.
```c++
SystemFile WriteToTextFile()
{
    // Open up normal SystemFile and do work with it
    SystemFile systemFile("file.txt", SystemFile::OpenMode::SF_OPEN_WRITE_ONLY);
    // ...
    // Need to interact with API which uses GenericStream?
    // Give ownership of the SystemFile to a SystemFileStream
    SystemFileStream systemFileStream(AZStd::move(systemFile));
    // Use AZ::Utils::SaveObjectToStream to serialize class to SystemFile
   AZ::Utils::SaveObjectToStream(systemFileStream, ...);

   // Give ownership of the created SystemFile back to the "systemFile" variable
   systemFile = AZStd::move(systemFileStream).MoveSystemFile();

   return systemFile;
}
```

**Comment by amzn-phist** [comment]
General comment: 
While I applaud the cleanup, looking at it now I don't really see much difference between `SystemFile` and `SystemFileStream`, other than the latter being an API wrapper/adapter.  Even so, the API differences aren't really that significant.  It makes me wonder if `SystemFile` should just implement the `GenericStream` API to begin with and fill in the differences, or some other reconfiguration of the two classes.  It's mostly a pass-through for what `SystemFile` already does, for instance in the `SystemFileStream` destructor is calling `Close()` but `SystemFile` already closes itself in its destructor.

**Comment by lumberyard-employee-dm** [comment]
> General comment: While I applaud the cleanup, looking at it now I don't really see much difference between `SystemFile` and `SystemFileStream`, other than the latter being an API wrapper/adapter. Even so, the API differences aren't really that significant. It makes me wonder if `SystemFile` should just implement the `GenericStream` API to begin with and fill in the differences, or some other reconfiguration of the two classes. It's mostly a pass-through for what `SystemFile` already does, for instance in the `SystemFileStream` destructor is calling `Close()` but `SystemFile` already closes itself in its destructor.

Well SystemFile is supposed to be a low level File API which doesn't use virtual functions to interact to read and write files on the file system. Furthermore it provides functionality for interacting with the filesystem.
It supports creating directories, searching for files, checking file metadata such as read-only or if the file is a directory.

The GenericStream API uses a virtual interface for working with a streaming API that other types IO mechanisms can in implement. It doesn't have any of the file system implementation functions.

File: Code/Framework/AzCore/AzCore/IO/GenericStreams.h (hunk 1 above)
Comment (amzn-phist) [reply Code/Framework/AzCore/AzCore/IO/GenericStreams.h:101]: Thanks!

**Review by cgalvan — APPROVED** [review]

**Review by moraaar — APPROVED** [review]
The change looks good to me. 

I guess SystemFileStream already has unit tests?
It'd be great to add new unit tests for the new ways to construct `SystemFileStream`.

In addition, with such fundamental core change I'd suggest to run basic tests in all platforms (maybe Lionbridge could help).

Lastly, will this affect VFS on Android? It'd be great if that could be tested as well (steps here https://github.com/moraaar/sig-platform/blob/Andoid_LKG_Workflow/testing-guidance/workflow-tests/android/android-lkg-workflows.md)

**Review by moraaar — COMMENTED** [review]
Thanks for adding the unit tests and great clean up as well!

**Comment by lumberyard-employee-dm** [comment]
> Thanks for adding the unit tests and great clean up as well!

No problem.

**Pull request merged**